OpenRA.Mods.Common/Traits/TunnelEntrance.cs

```diff
@@ -0,0 +1,65 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Linq;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class TunnelEntranceInfo : ITraitInfo, IRulesetLoaded
+	{
+		[Desc("Cell radius to use as a staging area for unit entry and exit.")]
+		public readonly int Margin = 4;
+
+		[Desc("Offset to check for the corresponding map tunnel portal.")]
+		public readonly CVec Sensor = CVec.Zero;
+
+		[Desc("Unit vector facing into tunnel.")]
+		public readonly CVec Direction = new CVec(1, 0);
+
+		public void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			if (Math.Abs(Direction.X) + Math.Abs(Direction.Y) != 1)
```

Comment:
This check is enforcing the assumptions that are made in the way `Direction` is used.  The way it searches for the matching tunnel portal would need to be completely redone if we wanted to relax this.